specifications/unified-provider.org

```diff
@@ -0,0 +1,84 @@
+has varied support for Snapcraft capabilities.
+
+This document defines an approach to refactor build provider
+=snapcraft.internal.build_providers._base_provider.Provider= such that it would
+support all current build providers (Host, Launchpad, LXD, Managed-Host,
+Multipass) and potential future providers (Remote LXD, etc.).
+
+* Provider Refactoring
+
+** Current Public Interfaces
+- ensure_provider()
+- setup_provider()
+- mount_project()
+- expose_prime()
+- execute_step()
+- pack_project()
+- clean(part_names)
+- clean_project()
+- shell()
+
+** Proposed Public Interfaces:
+
+|--------------------------------------------+---------------------------------------------------------------------------+-------------------------------------------------------|
+| method                                     | description                                                               | raises                                                |
+|--------------------------------------------+---------------------------------------------------------------------------+-------------------------------------------------------|
+| is_provider_ready() -> bool                | checks if provider is ready w/o installation or configuration             |                                                       |
+| setup_provider(echoer) -> None             | installs/configures provider, allowing for user prompting                 | ProviderInstallationError, ProviderConfigurationError |
+| execute_step(step: steps.Step) -> None     | executes (up to and including) step                                       | NotImplementedError                                   |
+| expose_prime(path: pathlib.Path) -> None   | mount project =prime= directory to host at specified path                 | NotImplementedError                                   |
+| snap() -> List[pathlib.Path]               | snap project (building as required), writing snap(s) to host project dir. |                                                       |
+| clean_parts(part_names: List[str]) -> None | clean project parts                                                       | NotImplementedError                                   |
+| clean() -> None                            | clean all project artifacts, including build-instance, if applicable.     |                                                       |
+| shell() -> None                            | launch interactive shell with build-instance, if applicable.              | NotImplementedError                                   |
+|--------------------------------------------+---------------------------------------------------------------------------+-------------------------------------------------------|
+
+*** Summary of changes:
+- Rename =ensure_provider()= to =is_provider_ready()=, scoping to see if
+  provider is ready without needing installation or further configuration.
+  Remove exceptions.  CLI will invoke setup_provider() if terminal is
```

Comment:
I don't understand this statement, can you make it clearer?

Reply:
will do

Reply:
should be done